transitions/loop.h

```diff
@@ -103,4 +103,69 @@ class TrLoopUntil {
   }
 };
 
+// Usage: TransitionLoopWhileL<LOOP_TRANSITION, END_TRANSITION, CONDITION>
+// LOOP_TRANSITION : TRANSITION
+// END_TRANSITION : TRANSITION
+// return value: COLOR
+//
+// Continuously runs LOOP_TRANSITION while CONDITION > 0
+// Makes more sense if TRANSITION is a TrConcat, as this will
+// transition to/from the intermediate steps in a loop.
+// If CONDITION <= 0, runs END_TRANSITION and stops.
+
+template<class LoopTr, class EndTr, class CONDITION>
+class TransitionLoopWhileL {
+
+public:
+void run(BladeBase* blade) {
+  condition_.run(blade);
+  if (!run_ || loop_tr_.done()) {
+    run_ = false;
+    if (condition_.calculate(blade)) {
+     run_ = true;
+     end_ = false;
+     loop_tr_.begin();
+    }
+  }
+  if (run_) {
+    if (!condition_.calculate(blade)) {
+      run_ = false;
+      end_ = true;
+      end_tr_.begin();
+    } else {
+      loop_tr_.run(blade);
+    }
+  }
+  if (end_) {
+    if (end_tr_.done()) {
+      end_ = false;
+    }
+    end_tr_.run(blade);
+    loop_tr_.run(blade);
+  }
+}
+  
+private:
+  bool run_ = false;
+  bool end_ = false;
+  LoopTr loop_tr_;
+  EndTr end_tr_;
+  PONUA SVFWrapper<CONDITION> condition_;
+
+public:
+auto getColor(int led) -> decltype(
+    MixColors(end_tr_.getColor(loop_tr_.getColor(RGBA_um_nod::Transparent(), RGBA_um_nod::Transparent(), led), RGBA_um_nod::Transparent(), led),
+        loop_tr_.getColor(loop_tr_.getColor(RGBA_um_nod::Transparent(), RGBA_um_nod::Transparent(), led), RGBA_um_nod::Transparent(), led), 1, 1)) {
+    if (run_) {
+      return loop_tr_.getColor(RGBA_um_nod::Transparent(),
+          RGBA_um_nod::Transparent(), led);
+    } else if (end_) {
```

Comment:
We shouldn't have a hard transition between "looping" and "end".
Instead, we should keep running the loop until "end" is done, and layer the end transition on top of the loop transition.

Reply:
Basically like:

decltype(SAME AS RETURN TYPE) ret = RGBA_um_nod::Transparent();
if (run_) {
  ret = loop_tr_.getColor(RGBA_um_nod::Transparent(), RGBA_um_nod::Transparent(), led);
  if (end_)  ret = loop_tr_.getColor(ret, RGBA_um_nod::Transparent(), led);
}
return ret;


The MixColors inside decltype() is just calculating the return type of the function, it doesn't do anything else.

Reply:
Thanks, I "think" I did it right BUT the end transition does seem to actually run, it just ends when the condition is false.  I tried moving the if (end_) outside the if(run_) but it did the same thing, so it must be something else?  Any ideas?